### PR TITLE
Json serialization

### PR DESCRIPTION
**Pull request by jsnmc**
Leverage the JavaScriptserializer to build up the json instead of building up the json from strings.
(Bug: 272033 Feature keys - handle special characters in description field)

### CODE REVIEW

File: src/AutoDeploy/DataCamel/Helpers/FileHelpers.cs [hunk 1]
```diff
@@ -66,36 +77,37 @@ public static void WriteLaunchKeysAsJson(string writeLocation)
 
         public static string ConvertToKeysfileJson(List<string> configs)
         {
```
Comment (jakebrown58) [inline src/AutoDeploy/DataCamel/Helpers/FileHelpers.cs]: As a general rule of thumb - I recommend an open-closed style approach.
Rather than re-implementing a function, create a different function, and leave the old one alone.   When you're satisfied with your new function, point the app to that function and remove the old one.

File: src/AutoDeploy/DataCamel/Helpers/FileHelpers.cs [hunk 2]
```diff
@@ -49,12 +51,21 @@ public static void Write(string fileName, List<string> s)
...
     {
-        public static List<string> GetLaunchKeys()
+        public static List<string> GetLaunchKeys(string config=@"C:\upgrade\AutoDeploy\volitleData.config")
         {
-            var userData = SimpleFileReader.Read(@"C:\upgrade\AutoDeploy\volitleData.config");
+            var userData = SimpleFileReader.Read(config);
             return userData.FindAll(x => x.StartsWith("LAUNCHKEY"));
         }
 
+        public static void WriteLaunchKeysAsJson(string writeLocation, string configLocation=null)
```
Comment (jakebrown58) [inline src/AutoDeploy/DataCamel/Helpers/FileHelpers.cs]: optional params are messy over time.
it would be better to use a provider pattern.  I can show you in a future commit what I mean, since that's probably not actionable advice.

File: src/AutoDeploy/DataCamel/Helpers/FileHelpers.cs [hunk 3]
```diff
@@ -49,12 +51,21 @@ public static void Write(string fileName, List<string> s)
 
     public class ConfigHelper
     {
-        public static List<string> GetLaunchKeys()
+        public static List<string> GetLaunchKeys(string config=@"C:\upgrade\AutoDeploy\volitleData.config")
```
Comment (jakebrown58) [inline src/AutoDeploy/DataCamel/Helpers/FileHelpers.cs]: should follow a provider pattern instead, as said elsewhere.

File: src/AutoDeploy/DataCamel/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -23,4 +23,5 @@
 [assembly: Guid("93d223dc-e2ea-4059-be2a-13016cbc6556")]
 
 [assembly: AssemblyVersion("1.4.6.0")]
-[assembly: AssemblyFileVersion("1.4.6.0")]
\ No newline at end of file
+[assembly: AssemblyFileVersion("1.4.6.0")]
+[assembly: InternalsVisibleTo("InstallerTests")]
```
Comment (jakebrown58) [inline src/AutoDeploy/DataCamel/Properties/AssemblyInfo.cs]: each of these apps is independent and shouldn't need to mess with internals, since none of these assemblies reference each other - and never should.    i'd rather keep things as public / private, and use interfaces for testing than using internals.

File: src/AutoDeploy/InstallerTests/DatabaseUpgraderTests/DatabaseUpgraderTests.cs [hunk 5]
```diff
@@ -42,7 +45,24 @@ public void DoIt()
         [TestMethod]
         public void WriteLaunchKeyJson()
         {
-            DataCamel.Helpers.ConfigHelper.WriteLaunchKeysAsJson(@"D:\test\testOutcome.json");
```
Comment (jakebrown58) [inline src/AutoDeploy/InstallerTests/DatabaseUpgraderTests/DatabaseUpgraderTests.cs]: whoops!   hard-paths in tests are evil.... thanks for getting rid of this one!   (My bad).

File: src/AutoDeploy/InstallerTests/DatabaseUpgraderTests/DatabaseUpgraderTests.cs [hunk 6]
```diff
@@ -42,7 +45,24 @@ public void DoIt()
...
+            JavaScriptSerializer serializer = new JavaScriptSerializer();
+            try
+            {
+                // Test for valid JSon
+                var obj = serializer.Deserialize<object>(keyFileContents);
+            }
+            catch (Exception ex)
+            {
+                Assert.Fail("JSon data could not be deserialized indicating bad data: " + ex.Message);
+            }
```
Comment (jakebrown58) [inline src/AutoDeploy/InstallerTests/DatabaseUpgraderTests/DatabaseUpgraderTests.cs]: ...something to consider for a follow-on change set would be to get rid of coupling this tests to the file system and relying on an embedded resource file in the solution.

this test should really have no coupling to the file system.    the in memory tests below should be adequate.

File: src/AutoDeploy/InstallerTests/DatabaseUpgraderTests/DatabaseUpgraderTests.cs [hunk 7]
```diff
@@ -54,7 +74,23 @@ public void ConvertLaunchKeyConfigToLaunchKeyJson()
...
+        {
+            List<string> configs = new List<string>();
+            configs.Add("LAUNCHKEY|MyKey=\"someFeature\" with -- some \"special\" chars");
+            configs.Add("LAUNCHKEY|MyKey2=\"someFeature2\"");
+            var x = DataCamel.Helpers.ConfigHelper.ConvertToKeysfileJson(configs);
+            Console.WriteLine(x);
+
+            var targetMatch = "[{\"Description\":\"someFeature\\\" with -- some \\\"special\\\" char\",\"FeatureKey\":\"MyKey\"},{\"Description\":\"someFeature2\",\"FeatureKey\":\"MyKey2\"}]";
+            Assert.AreEqual(x, targetMatch);
         }
```
Comment (jakebrown58) [inline src/AutoDeploy/InstallerTests/DatabaseUpgraderTests/DatabaseUpgraderTests.cs]: nice.

File: src/AutoDeploy/InstallerTests/InstallerTests.csproj [hunk 8]
```diff
@@ -141,6 +142,9 @@
     </ProjectReference>
   </ItemGroup>
   <ItemGroup>
+    <None Include="DatabaseUpgraderTests\testData.config">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
```
Comment (jakebrown58) [inline src/AutoDeploy/InstallerTests/InstallerTests.csproj]: as said above, ideally we can rip this out rather than bloating our test project with resource files that are easily generated in memory.

File: src/InstallerService/InstallerService/Daemon/Controllers/AvailableFeaturesController.cs [hunk 9]
```diff
@@ -145,27 +145,31 @@ public HttpResponseMessage GetKeys(string dropLocation)
...
-                if (!dropLocation.EndsWith(@"\"))
+                if (!dropLocation.StartsWith("\\"))
                 {
-                    dropLocation = dropLocation + @"\";
+                    // open volitle data.
+                    var volitleData = FileHelpers.ReadConfigAsData("volitleData.config");
+                    var folderRoot = volitleData.Find(x => x.Contains("BUILD_FOLDER_ROOT"));
+                    if (string.IsNullOrEmpty(folderRoot))
+                    {
+                        throw new Exception("BUILD_FOLDER_ROOT not set in config");
```
Comment (jakebrown58) [inline src/InstallerService/InstallerService/Daemon/Controllers/AvailableFeaturesController.cs]: What http response does this give if we throw an exception?
Will this response be understandable to the consumer?

Would be smart to split this logic into a helper and add tests, as this code is pretty squirrely.  (It's a complicated mess - I wrote it - Without Tests.    !Sorry!)

File: src/AutoDeploy/InstallerTests/DatabaseUpgraderTests/DatabaseUpgraderTests.cs (hunk 6 above)
Comment (jsnmc) [inline src/AutoDeploy/InstallerTests/DatabaseUpgraderTests/DatabaseUpgraderTests.cs]: I made this change as it isn't a big change.  My thought with using the embedded resource was to work with data that is more easily understood.

So in the file I had the data like this:
`LAUNCHKEY|MyKey="someFeature with "some" "`

In the code, I have to define it like this:
`"LAUNCHKEY|MyKey=\"someFeature with \"some\"\""`

Change made, not a big deal.

File: src/AutoDeploy/DataCamel/Properties/AssemblyInfo.cs (hunk 4 above)
Comment (jsnmc) [inline src/AutoDeploy/DataCamel/Properties/AssemblyInfo.cs]: I realized that I didn't need this.  Initially I made the config file paths, internal to be changed only by the test.  I changed that to optional params ( which I will most likely change ) to avoid it.

File: src/AutoDeploy/InstallerTests/DatabaseUpgraderTests/DatabaseUpgraderTests.cs (hunk 6 above)
Comment (jakebrown58) [inline src/AutoDeploy/InstallerTests/DatabaseUpgraderTests/DatabaseUpgraderTests.cs]: yeah, it's definately a tradeoff.   i like the clarity of a raw file where you don't have to escape characters.    ....but i like not having a dependency on the file system in the unit tests, maybe just slightly more.

File: src/InstallerService/InstallerService/Daemon/Controllers/AvailableFeaturesController.cs (hunk 9 above)
Comment (jsnmc) [inline src/InstallerService/InstallerService/Daemon/Controllers/AvailableFeaturesController.cs]: Not sure why this change seems to show so many diffs in Github.  Essentially what I did was elevate the try to the start of the logic.  There is a catch below that was already handling reporting to the http side.  I just wanted a specific error message as I hit the original cryptic message a few times in development.  So it literally is just to report "BUILD_FOLDER_ROOT not set in config"
